src/engine/enginemaster.h

```diff
@@ -126,68 +128,104 @@ class EngineMaster : public QObject, public AudioSource {
+        bool m_fadeout;
     };
 
     class GainCalculator {
       public:
         virtual double getGain(ChannelInfo* pChannelInfo) const = 0;
     };
-    class ConstantGainCalculator : public GainCalculator {
+    class PflGainCalculator : public GainCalculator {
       public:
         inline double getGain(ChannelInfo* pChannelInfo) const {
-            return pChannelInfo->m_pChannel->isTalkover() ? m_dTalkoverGain : m_dGain;
+            Q_UNUSED(pChannelInfo);
+            return m_dGain;
         }
         inline void setGain(double dGain) {
             m_dGain = dGain;
         }
-        inline void setTalkoverGain(double dGain) {
-            m_dTalkoverGain = dGain;
-        }
       private:
         double m_dGain;
-        double m_dTalkoverGain;
+    };
+    class TalkoverGainCalculator : public GainCalculator {
+      public:
+        inline double getGain(ChannelInfo* pChannelInfo) const {
+            Q_UNUSED(pChannelInfo);
+            return 1.0;
+        }
     };
     class OrientationVolumeGainCalculator : public GainCalculator {
       public:
         OrientationVolumeGainCalculator()
-                : m_dVolume(1.0), m_dLeftGain(1.0), m_dCenterGain(1.0), m_dRightGain(1.0),
-                  m_dTalkoverGain(1.0) { }
+                : m_dVolume(1.0),
+                  m_dLeftGain(1.0),
+                  m_dCenterGain(1.0),
+                  m_dRightGain(1.0) {
+        }
 
         inline double getGain(ChannelInfo* pChannelInfo) const {
-            if (pChannelInfo->m_pMuteControl->get() > 0.0) {
-                return 0.0;
-            }
-            if (pChannelInfo->m_pChannel->isTalkover()) {
-                return m_dTalkoverGain;
-            }
             const double channelVolume = pChannelInfo->m_pVolumeControl->get();
             const double orientationGain = EngineMaster::gainForOrientation(
-                pChannelInfo->m_pChannel->getOrientation(),
-                m_dLeftGain, m_dCenterGain, m_dRightGain);
+                    pChannelInfo->m_pChannel->getOrientation(),
+                    m_dLeftGain, m_dCenterGain, m_dRightGain);
             return m_dVolume * channelVolume * orientationGain;
         }
 
-        inline void setGains(double dVolume, double leftGain, double centerGain, double rightGain,
-                             double talkoverGain) {
+        inline void setGains(double dVolume, double leftGain,
+                double centerGain, double rightGain) {
             m_dVolume = dVolume;
             m_dLeftGain = leftGain;
             m_dCenterGain = centerGain;
             m_dRightGain = rightGain;
-            m_dTalkoverGain = talkoverGain;
         }
 
       private:
-        double m_dVolume, m_dLeftGain, m_dCenterGain, m_dRightGain, m_dTalkoverGain;
+        double m_dVolume;
+        double m_dLeftGain;
+        double m_dCenterGain;
+        double m_dRightGain;
+    };
+    template<typename T, unsigned int CAPACITY>
+    class FastVector {
```

Comment:
> Can you point me to the place where it is done? Have not found it.

Iterators? Hm, it looks like they were added in 4.8.0:
https://qt.gitorious.org/qt/qt/source/8ab1ad64620ff9d0453a326010d161ea68a63a2f:src/corelib/tools/qvarlengtharray.h#L154
https://bugreports.qt.io/browse/QTBUG-14010
so we shouldn't use them. You could access the ```data()``` pointer directly though to do unchecked access.

> We cannot exceed kMaxChannels because the sample util fictions are not prepared for a dynamic size. 
> I also think there is no need for this and a fixed limit will suit here.

Hm, yes you're right. I remember writing a loop fallback for >32 channels but it seems to be gone now.

> If would make more channels than kMaxChannels work and switch to QVarLengthArray, we will introduce malloc calls in every engine thread call which is also not desirable.

This is only because you declare the FastVectors on the stack in this PR. If you declared them as a member variable and clear()'d then that would only require them to malloc once. 

Malloc'ing in the callback is also better than appearing broken.

> So lets keep the FastVector for now.

I'd rather not... why do we need it?

It still doesn't work properly for complex types. If the type is complex, append should run the in-place copy constructor as QVarLengthArray does: https://qt.gitorious.org/qt/qt/source/8ab1ad64620ff9d0453a326010d161ea68a63a2f:src/corelib/tools/qvarlengtharray.h#L123

QVarLengthArray is better in pretty much all respects. Why re-invent the wheel?